modules/auth.php

```diff
@@ -260,6 +262,94 @@ function handler_admin_authgroupesx($page, $action = 'list', $id = null)
         $table_editor->describe('last_used', 'dernière utilisation', true);
         $table_editor->apply($page, $action, $id);
     }
+
+    /** Handles the Discourse SSO authentication.
+     * https://meta.discourse.org/t/official-single-sign-on-for-discourse/13045
+     * Expects the Discourse domain as the last component of the URL path
+     * Expects the following GET parameters: sso, sig
+     * e.g. https://www.polytechnique.org/auth-discourse/forum.polytechnique.org?sso=...&sig=...
+     */
+    function handler_discourse($page, $domain = '')
+    {
+        global $globals;
+
+        // Load the key
+        if (!preg_match('/^[a-zA-Z0-9\-.:]+$/', $domain)) {
+            $page->kill("Domaine non valide");
+        } elseif (empty($globals->discourse->$domain)) {
+            $page->kill("Domaine inconnu");
+        }
+        $ext_url = 'https://' . $domain. '/session/sso_login';
+        $key = $globals->discourse->$domain;
+
+        // Check the signature of the given nonce
+        $sso_data_b64 = Get::s('sso');
+        $sign = hash_hmac('sha256', $sso_data_b64, $key);
+        if (!secure_string_compare($sign, Get::s('sig'))) {
+            $page->kill("Signature invalide");
+        }
+        $sso_data = array();
+
+        // Decode the SSO data, which must contain two fields:
+        // * nonce: a nonce which has to be given back in the SSO response
+        // * return_sso_url: the return URL, which should be $ext_url
+        parse_str(base64_decode($sso_data_b64), $sso_data);
+        if (empty($sso_data['nonce'])) {
+            $page->kill("Données SSO non valides");
+        }
+        // return_sso_url is allowed to be either HTTP or HTTPS, even though we are forcing an HTTPS response
```

Comment:
The official documentation uses HTTP to build the return URL (out of nowhere) and I don't know whether our Discourse instance will fill ``return_sso_url`` field with ``https://`` instead of ``http://``. So this test accepts both but returns on https. Once the system is live, if ``return_sso_url`` uses ``https://``, it would be possible to simplify this test.